synapse/push/bulk_push_rule_evaluator.py

```diff
@@ -196,125 +210,139 @@ async def _get_power_levels_and_sender_level(
 
     @measure_func("action_for_event_by_user")
     async def action_for_event_by_user(
-        self, event: EventBase, context: EventContext
+        self, events_and_context: List[Tuple[EventBase, EventContext]]
     ) -> None:
-        """Given an event and context, evaluate the push rules, check if the message
-        should increment the unread count, and insert the results into the
-        event_push_actions_staging table.
+        """Given a list of events and their associated contexts, evaluate the push rules
+        for each event, check if the message should increment the unread count, and
+        insert the results into the event_push_actions_staging table.
         """
-        if not event.internal_metadata.is_notifiable():
-            # Push rules for events that aren't notifiable can't be processed by this
-            return
-
-        # Disable counting as unread unless the experimental configuration is
-        # enabled, as it can cause additional (unwanted) rows to be added to the
-        # event_push_actions table.
-        count_as_unread = False
-        if self.hs.config.experimental.msc2654_enabled:
-            count_as_unread = _should_count_as_unread(event, context)
-
-        rules_by_user = await self._get_rules_for_event(event)
-        actions_by_user: Dict[str, Collection[Union[Mapping, str]]] = {}
-
-        room_member_count = await self.store.get_number_joined_users_in_room(
-            event.room_id
-        )
+        for event, context in events_and_context:
+            if not event.internal_metadata.is_notifiable():
+                # Push rules for events that aren't notifiable can't be processed by this
+                return
+            # Skip push notification actions for historical messages
+            # because we don't want to notify people about old history back in time.
+            # The historical messages also do not have the proper `context.current_state_ids`
+            # and `state_groups` because they have `prev_events` that aren't persisted yet
+            # (historical messages persisted in reverse-chronological order).
+            if event.internal_metadata.is_historical():
+                return
```

Comment:
Note to self, this bit about historical messages was moved here from one of the callers.

Was it a bug that the other callers didn't have this check?

Reply:
That's a great question and looking at the other callsites they are all related to handling events coming in over federation, so if historical events were coming in over federation than perhaps? I suppose the mechanism for that would be: historical events are persisted in one room in originating server -> sent out over federation to other servers, but I am not super clear on how historical messages are handled (i.e. do they keep the "historical" designation after being persisted on the first server?)

Reply:
I think this might refer to MSC2716? @MadLittleMods Any ideas?

Reply:
Is `event.internal_metadata.is_notifiable()` new? It seems like we should include `event.internal_metadata.is_historical()` in that check. And then we don't need the extra condition here.

The historical designation only lives on the origin server. One of the goals is to inhibit push notifications.

Historical events have a negative `stream_ordering` because they are marked as backfilled so they don't go down `/sync` either.

For historical events coming over federation, they are coming over `/backfill` which also means they get a negative `stream_ordering` on the remote server and don't come down `/sync` or push notifications.

Reply:
> Is `event.internal_metadata.is_notifiable()` new?

Yes, it was added in #13719.

> For reference, I don't know what "other callers" is referring to.

They're mostly callers for persisting events from federation. It sounds like those events will never have `is_historical()` resolve to `True`, but still good to consolidate code paths?

Reply:
> They're mostly callers for persisting events from federation. It sounds like those events will never have `is_historical()` resolve to `True`, but still good to consolidate code paths?

Seems good to me and doesn't change current functionality for MSC2716 historical things 👍

Probably best to do the `event.internal_metadata.is_notifiable()` consolidation further ⏩

Reply:
Thanks for the insights @MadLittleMods! Maybe @H-Shay can do that as a follow-up? 😄

Reply:
Ya no problem!